zcash_primitives/src/merkle_tree.rs

```diff
@@ -44,590 +23,307 @@ pub trait HashSer {
+}
 
-    fn append_inner(&mut self, node: Node, depth: usize) -> Result<(), ()> {
-        if let Some(mut cursor) = self.cursor.take() {
-            cursor
-                .append_inner(node, depth)
-                .expect("cursor should not be full");
-            if cursor.is_complete(self.cursor_depth) {
-                self.filled
-                    .push(cursor.root_inner(self.cursor_depth, PathFiller::empty()));
-            } else {
-                self.cursor = Some(cursor);
-            }
-        } else {
-            self.cursor_depth = self.next_depth();
-            if self.cursor_depth >= depth {
-                // Tree is full
-                return Err(());
-            }
+/// Serializes an `IncrementalWitness` as an array of bytes.
+pub fn write_incremental_witness<Node: HashSer, W: Write, const DEPTH: u8>(
+    witness: &IncrementalWitness<Node, DEPTH>,
+    mut writer: W,
+) -> io::Result<()> {
+    write_commitment_tree(witness.tree(), &mut writer)?;
+    Vector::write(&mut writer, witness.filled(), |w, n| n.write(w))?;
+    Optional::write(&mut writer, witness.cursor().as_ref(), |w, t| {
+        write_commitment_tree(t, w)
+    })
+}
 
-            if self.cursor_depth == 0 {
-                self.filled.push(node);
+/// Reads a Merkle path from its serialized form.
+pub fn merkle_path_from_slice<Node: HashSer, const DEPTH: u8>(
+    mut witness: &[u8],
+) -> Result<MerklePath<Node, DEPTH>, ()> {
+    // Skip the first byte, which should be DEPTH to signify the length of
+    // the following vector of Pedersen hashes.
+    if witness[0] != DEPTH {
+        return Err(());
+    }
+    witness = &witness[1..];
+
+    // Begin to construct the authentication path
+    let iter = witness.chunks_exact(33);
+    witness = iter.remainder();
+
+    // The vector works in reverse
+    let auth_path = iter
+        .rev()
+        .map(|bytes| {
+            // Length of inner vector should be the length of a Pedersen hash
+            if bytes[0] == 32 {
+                // Sibling node should be an element of Fr
+                Node::read(&bytes[1..]).map_err(|_| ())
             } else {
-                let mut cursor = CommitmentTree::empty();
-                cursor
-                    .append_inner(node, depth)
-                    .expect("cursor should not be full");
-                self.cursor = Some(cursor);
+                Err(())
             }
-        }
-
-        Ok(())
+        })
+        .collect::<Result<Vec<_>, _>>()?;
+    if auth_path.len() != usize::from(DEPTH) {
+        return Err(());
     }
 
-    /// Returns the current root of the tree corresponding to the witness.
-    pub fn root(&self) -> Node {
-        self.root_inner(SAPLING_COMMITMENT_TREE_DEPTH)
-    }
+    // Read the position from the witness
+    let position = witness
+        .read_u64::<LittleEndian>()
+        .map_err(|_| ())
+        .and_then(|p| Position::try_from(p).map_err(|_| ()))?;
```

Comment:
```suggestion
        .and_then(|p| Position::try_from(p))?;
```
(I think this is right.)